RFC: priority-based mempool (#154)
tendermint/spec: rfc/004-priority-based-mempool.md

To make Tendermint support priority-based transaction processing, two main aspects of current architecture in Tendermint need to be updated: one is on ABCI's interface for accepting a transaction (`CheckTx`), and another is on Tendermint's mempool design and implementation. This RFC proposes corresponding changes to address these two aspects.

## Proposal

Add a new field `priority` to the ABCI `ResponseCheckTx` message:

```proto
message ResponseCheckTx {
  uint32         code       = 1;
  bytes          data       = 2;
  string         log        = 3;  // nondeterministic
  string         info       = 4;  // nondeterministic
  int64          gas_wanted = 5 [json_name = "gas_wanted"];
  int64          gas_used   = 6 [json_name = "gas_used"];
  repeated Event events     = 7
      [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
  string codespace = 8;
  uint64 priority  = 9;
}
```

> Then ABCI developers can add application-specific user-defined priorities for the transactions. This addition is naturally backward-compatible because by default all requests will be returned with 0 priority and on mempool side the process will fallback to FIFO.

Reviewer: Does this imply that all txs with the same priority will be processed in FIFO order? If so, that places an additional constraint on any sorting or heap algorithm, i.e. that it is stable.

Author: agree, but easy tweaks can be made to achieve that even though the sorting is not stable: e.g. using "priority + timestamp" as the comparing key such that if two items have the same priority, the one with lower timestamp (earlier) should be ranked higher.

The next step is to make current mempool implementation priority-aware. We can have 3 phases to gradually implement this feature to make as least breaking changes as possible.

### Phase 0: Keep existing CList-based mempool

Goal:

1. Zero cost on processing "legacy" transactions (i.e. those have 0 priority)
2. When creating a block, transactions with higher priorities should be included earlier

The first phase will be experimental and only achieves the desired functionality *with potential performance degradation*. The main purpose is to design the code flow and underlying APIs accordingly, without dramatic code changes on existing data structures / algorithms.

> The simplest approach I can think of is:

Reviewer 2: ```suggestion
A simple approach we can take is the following:
```

1. When reading `CheckTx` responses, mempool keeps a single counter called `maxPriority`. If all transactions have 0 priorities, the counter should simply have the same 0 value
2. When `Mempool.ReapMaxBytesMaxGas` is called by consensus engine, mempool would check its `maxPriority` value:
    - if it's zero, it should follow existing logic to return eligible transactions in FIFO style

>     - if not, should return eligible transactions ranked by their priorities (either by `O(n^2)` brute-force or `O(nlogn)` sort). Then during commit, update mempool's `maxPriority` accordingly since transactions will be removed from mempool

Reviewer 3: not sure how viable it is to sort mempool after every block

Reviewer 2: It should be nlogn really. What's the typical size of the mempool?

Reviewer 4: I think the default is 5000 txs

Reviewer 5: The mempool size is currently a function of the software being unable to handle more. Ideally it should further scale to demand, especially as more L1 scaling solutions land. 

For ref, ETH per txstreet.com is >75k txs in mempool.

Author: agree O(n^2) brute force should be a no-go. will update

    // Exclusive mutex for Update method to prevent concurrent execution of
    // CheckTx or ReapMaxBytesMaxGas(ReapMaxTxs) methods.
    updateMtx sync.RWMutex

    wal          *auto.AutoFile // a log of mempool txs
    proxyAppConn proxy.AppConnMempool

    logger  log.Logger
    metrics *Metrics
}

type mempoolImpl interface {
    Size() int

    addTx(*mempoolTx, uint64)
    removeTx(types.Tx) bool // return whether corresponding element is removed or not
    updateRecheckCursor()
    reapMaxTxs(int) types.Txs
    reapMaxBytesMaxGas(int64, int64) types.Txs // based on priority
    recheckTxs(proxy.AppConnMempool)
    isRecheckCursorNil() bool
    getRecheckCursorTx() *mempoolTx
    getMempoolTx(types.Tx) *mempoolTx
    deleteAll()
    // ...and more
}
```

Such that mempool's shared code will live under `basemempool` (callback handling, WAL, etc.) while different `mempoolImpl` only needs to implement required methods on transaction addition / removal / iteration etc. The proposed interface is [implemented in this code base](https://github.com/QuarkChain/tendermintx/blob/master/mempool/mempool.go).

### Phase 2: Implement mempool based on different data structures

> We will have more discussions on what data structure fits the requirements best. For reference, we have some preliminary profiling results on [left-leaning red–black tree and BTree](https://github.com/QuarkChain/tendermintx/blob/master/mempool/bench_test.go#L18-L31).

Reviewer 2: Why do we need anything fancy here? Can we not just use a max heap?

Reviewer 5: You want a data structure that can be added to in parallel in a lock-free fashion. Neither max heaps or RB trees satisfy this to my knowledge. 

I am still in favor of the concurrent lock-free AVL variant from: http://ddana.cswp.cs.technion.ac.il/wp-content/uploads/sites/19/2015/12/logicalorderingavl.pdf

ref: https://github.com/tendermint/tendermint/issues/2147#issue-347673061

Author: 1. using balance tree can make the iteration during reaping mempool O(n) instead of O(nlogn) as it’s just an in-order traversal, while cost would be amortized during insertions. essentially a trade-off between performances of *tx addition* and *reaping*
1. a lock-free data structure supporting parallel addition sounds really interesting. but considering the trade-off between performance gain and implementation complexity, I think some profilings can help get more insights. I would say if a simpler lock-based data structure can provide similar performance as current `CList`  then it’s ok to have a lock

Reviewer 3: > I am strongly in favor of the concurrent lock-free AVL variant from: http://ddana.cswp.cs.technion.ac.il/wp-content/uploads/sites/19/2015/12/logicalorderingavl.pdf

are there any implementations?

Reviewer 3: > a lock-free data structure supporting parallel addition sounds really interesting.

yeah. also, being able to Reap transactions from the mempool and, in the same time, iterate over the tree (broadcasting transactions) would be nice too.

Reviewer: Maybe I'm wrong here, but aren't `CheckTx` calls synchronous and serialized? So why do we need a lock-free data structure when the access to it will be serialized by `CheckTx` anyway?

Reviewer 3: > Maybe I'm wrong here, but aren't CheckTx calls synchronous and serialized?

they're supposed to be async https://github.com/tendermint/tendermint/blob/d1a88fe39f0200b89b406d783a5a2490cfe3c42a/mempool/clist_mempool.go#L291

Reviewer 6: > Maybe I'm wrong here, but aren't CheckTx calls synchronous and serialized? So why do we need a lock-free data structure when the access to it will be serialized by CheckTx anyway?

The calls are made asynchronously, but in serial order. But we may want to perform two other operations simultaneously:
- broadcast txs to peers (requires reading from the set of txs)
- reaping txs for a proposal block (requires reading an ordered batch of txs)

![mempool2](https://user-images.githubusercontent.com/1282182/92575388-52b91200-f299-11ea-91c8-10e05e7be4df.png)

It’s been reported that the linked list and the map could [diverge after the cache becomes full](https://github.com/tendermint/tendermint/issues/5281).

The biggest downside of the current implementation is the lack of prioritization. It should give the ABCI application a way to establish some order, based on the fee or some other field(s).

Dev and others also expressed some concerns about concurrency patters - “My concern is primarily that we currently have the downside that you can't keep receiving / broadcasting txs while updating / rechecking.“ (see [this issue](https://github.com/tendermint/tendermint/issues/2147) and [this one](https://github.com/tendermint/tendermint/issues/2484))

Other possible optimizations:

- [highwayhash instead of SHA256](https://github.com/tendermint/tendermint/issues/2187)
- [remove storing height](https://github.com/tendermint/tendermint/issues/2834)
- [remove gasWanted for non-validators](https://github.com/tendermint/tendermint/issues/2978)
- [time-based eviction](https://github.com/tendermint/tendermint/issues/3436)

> What we want from the new design:

Reviewer 6: I think it would be helpful to have a clear more comprehensive statement of the functioning of the mempool. It should include the need to:

- Broadcast txs (in order) to peers, concurrently and efficiently (ie. without resending the same txs many times)
- Execute CheckTx serially but concurrently with other Tendermint functionality (like executing a block)
- Allow proposers to efficiently reap txs from the mempool (possibly concurrently with new CheckTxs happening?)
- Support tx rechecking and re-prioritization (?!) after a block is executed 
- Ensure priority ordered txs still form a valid ordering (ie. they pass CheckTx in that order and thus would pass DeliverTx in that order)
- 
- More?

The mempool has to handle a diverse range of concerns and it would be good to get clarity on them upfront so we can weigh the full set of constraints when evaluating designs. 

It also might give us an opportunity to realize that certain requirements are actually not necessary, or certain other requirements are necessary.

Reviewer 3: > Ensure priority ordered txs still form a valid ordering (ie. they pass CheckTx in that order and thus would pass DeliverTx in that order)

Does this mean we'll have to re-execute CheckTx on the whole mempool every time we sort it? Otherwise, how would we guarantee that a sequence will pass?

> re-prioritization (?!)

we won't be needing this if we choose a tree, which is always sorted (actual)

Reviewer 6: > Does this mean we'll have to re-execute CheckTx on the whole mempool every time we sort it? Otherwise, how would we guarantee that a sequence will pass?

This is going to be a challenge. We'll have to figure something out here. Possibly the mempool will have to be more aware of the relationship between txs (eg if they pertain to the same "account", whatever that might be from tendermint's perspective). One way to possibly do this is leave it to the app to return a list of txs that are invalidated by a new tx, for instance ... (?)

Reviewer 3: > Ensure priority ordered txs still form a valid ordering (ie. they pass CheckTx in that order and thus would pass DeliverTx in that order)

where is this requirement coming from?

Reviewer 6: From the fact that we need proposers to proposer blocks that have a sequence of validly ordered txs. Like if I send a gaia tx with sequence 5 and a small fee, and then a nother tx with sequence 6 and a large fee, the sequence 5 tx still needs to come first! (in bitcoin land they might call this "child pays for parent")

Reviewer 3: Right. Then we might need that "account" idea when `CheckTx` optionally returns a grouping key ("account adress") and a sequence number. The proposer would need to replace a transaction with lower sequence number if such exists.

### Phase 0: Keep existing CList-based mempool

Goal:

1. Zero cost on processing "legacy" transactions (i.e. those have 0 priority)
2. When creating a block, transactions with higher priorities should be included earlier

The first phase will be experimental and only achieves the desired functionality *with potential performance degradation*. The main purpose is to design the code flow and underlying APIs accordingly, without dramatic code changes on existing data structures / algorithms.

A simple approach we can take is the following:

1. When reading `CheckTx` responses, mempool keeps a single counter called `maxPriority`. If all transactions have 0 priorities, the counter should simply have the same 0 value
2. When `Mempool.ReapMaxBytesMaxGas` is called by consensus engine, mempool would check its `maxPriority` value:
    - if it's zero, it should follow existing logic to return eligible transactions in FIFO style
    - if not, should return eligible transactions ranked by their priorities (e.g. `O(nlogn)` sort). Then during commit, update mempool's `maxPriority` accordingly since transactions will be removed from mempool

> ### Phase 1: Abstract away existing mempool interface

Reviewer 6: Seems like a great idea and should move forward independently of the rest of this RFC - probably best to move this part to an ADR in Tendermint directly (it doesn't break protocols or APIs so no need for spec-based RFC process, just changes to Tendermint Go I think) and work on it there while we figure out how to actually support the protocol changes for prioritization

Author: sounds good. I'll draft an ADR covering details the refactoring.

Author: update: ADR proposed in https://github.com/tendermint/tendermint/pull/5451

### Concurrency Issues

> Dev and others also expressed some concerns about concurrency patters - “My concern is primarily that we currently have the downside that you can't keep receiving / broadcasting txs while updating / rechecking.“ (see [this issue](https://github.com/tendermint/tendermint/issues/2147) and [this one](https://github.com/tendermint/tendermint/issues/2484))

A smaller tree can be used to receive transactions and broadcast them to other peers. This tree will be merged into the main one once rechecking is done.

> “I'm not sure that that concurrent AVL tree is sufficient for determinism...” Jae

When fee is always zero (for those applications who don’t use fee), it may be possible to issue sequence numbers and use them to sort transactions.

## Status

Proposed

## Consequences

### Positive

- ABCI blockchains will have more customizability on how transactions are included in blocks

### Negative

- At early phases, using priority-based mempools could face performance degradations
- Mempool implementations may have increased complexities

### Neutral

- Mempool code could undergo a non-trivial refactoring

> ## Appendix A. Other Projects' Memory Pools

Reviewer 6: Would be helpful to provide some summarization here of how both the Libra and GoEthereum mempools work for context. Likely we can borrow quite a bit from them.

Reviewer 3: libra mempool maintains a hashmap of accounts (each account contains a list of transactions, sorted by sequence number). Btree is used for `PriorityIndex` (global index of transactions sorted by priority).

Reviewer 6: Right. Let's add that to the text. Maybe we need ResponseCheckTx to include an "address" field so txs can be grouped according to their "sender" (whatever that is from tendermint's perspective lol) which will help manage some of the complexity here. Iirc, ethereum does something similar.

This is a case where being agnostic to txs is going to be challenging. Either we leak more through the abci boundary or we give the app more control here.

Author: that's exactly the problem we faced when trying to simulate the "nonce" system from ethereum. implementing something like that could be useful but I think it's beyond the scope of this RFC and deserves its own.

will update the texts on how it's done in libra and eth.
